File: readErrorMap.m
function error_map = readErrorMap(filepath,numOfLines, graph_type)
% error_map = READERRORMAP(s) Read the error map test file and return the map as a matrix
% The function go through the file, check it's consistency and start to parse the data
% The function then saves everything in 'error_map' and returns it.
% s		The file to read

%lines = countLines(s);
fid = fopen(filepath);
meta = metaData(str2num(fgets(fid)));
% skip the "pages order" line
fgets(fid);

if(meta.testID ~= testID.errorMap)
     err = sprintf('Incompatible file parser.\n trying to use readErrorMAp file parser while metaData specifies test ID of %d',...
            meta.testID);
        msgbox(err,'Error in test ID of wrong file parser');
else
    wb = waitbar(0,sprintf('Reading Error Map: %d%%',0)); 
    m = zeros(meta.pagesPerBlock,meta.bytesPerPage*8);
    iter = 0;
    while ~feof(fid)
       arr =  str2num(fgets(fid));
       i = 2;
       while i < length(arr)
          page = arr(i);
          i = i+1;
          if arr(i) ~= 0
            bits = arr(i+1:i+arr(i));
          else
              bits = [];
          end
          i = i+arr(i)+1;
          if arr(i) ~= 0
            bits = [bits,arr(i+1:i+arr(i))];
          end
          i = i+arr(i)+1;
          m(page+1,bits+1) = m(page+1,bits+1)+1;      
       end
       iter = iter + 1;
       waitbar(double(iter/numOfLines),wb,sprintf('Reading Error Map: P/E Cycle %d/%d',iter,numOfLines));
    end
    close(wb);
    delete(wb);
end
fclose(fid);
pages_order = pagesOrder(filepath);
ppb = meta.pagesPerBlock;

% calculate pages order
if ( meta.architecture == architecture.mlc)
    if strcmp(graph_type, 'Levels')
        left_low = m(pages_order(1:2:ppb/2)+1,:);
        right_low = m(pages_order(2:2:ppb/2)+1,:);
        left_high = m(pages_order(ppb/2+1:2:end)+1,:);
        right_high = m(pages_order(ppb/2+2:2:end)+1,:);
        left = [left_low;left_high];
        right = [right_low;right_high];
    else
        low_pages = pages_order(1:ppb/2);
        high_pages = pages_order((ppb/2) + 1:end);

        sums = zeros(ppb/2,meta.bytesPerPage*8);
        for i = 1:ppb/2
            %asuming an error in both pages is not possible. 
            sums(i,:) = m(low_pages(i)+1,:)+m(high_pages(i)+1,:);
        end
        left = sums(1:2:end,:);
        right = sums(2:2:end,:);
    end
 
elseif (meta.architecture == architecture.tlc)
    if strcmp(graph_type, 'Levels')
        left_low = m(pages_order(1:2:ppb/3)+1,:);
        right_low = m(pages_order(2:2:ppb/3)+1,:);
        left_mid = m(pages_order(ppb/3+1:2:(2*(ppb/3)))+1,:);
        right_mid = m(pages_order(ppb/3+2:2:(2*(ppb/3)))+1,:);
        left_high = m(pages_order(2*(ppb/3)+1:2:end)+1,:);
        right_high = m(pages_order(2*(ppb/3)+2:2:end)+1,:);
        
        left = [left_low; left_mid; left_high];
        right = [right_low; right_mid; right_high];
    else
        sums = zeros(ppb/3, meta.bytesPerPage*8);
        triplets = [pages_order(1:ppb/3);
                    pages_order(ppb/3+1:2*(ppb/3));
                    pages_order(2*(ppb/3)+1:end)];

        for triplet = 1:ppb/3
            sums(triplet,:) = sum(m(triplets(:,triplet)+1,:));
        end
        left = sums(1:2:ppb/3,:);
        right = sums(2:2:ppb/3,:);
    end
    
else
    err = sprintf('Unsupported architecture for BitErrorMap.');
    msgbox(err,'Unsupported manufacturer'); 
    return;
end

error_map = [left,right];

File: readErrorMap.py
import numpy as np
from numpy import zeros, concatenate, fromstring, uint32
import os
import hashlib
import time


READ_CHUNK_SIZE = 300*1024*1024
SUPPORTED_READ_MODES = ['normal', 'levels']

def calc_sha1(filepath):
	t1 = time.clock()
	_hash = hashlib.sha1()
	with open(filepath, 'rb') as f:
		while True:
			# we use the read passing the size of the block to avoid
			# heavy ram usage
			data = f.read(READ_CHUNK_SIZE)
			if not data:
				break
			# calculate partial hash (so far)
			_hash.update(data)
	return _hash.hexdigest()

	
def read_error_map(filepath, read_mode='normal', save_path=None):
	read_mode = read_mode.lower()
	print(read_mode)
	if read_mode not in SUPPORTED_READ_MODES:
		print('read_mode must be on of the following modes: {}. Got {}.'
				.format(' | '.join(SUPPORTED_READ_MODES), read_mode))
		exit(-1)
	
	path, filename = os.path.split(filepath)
	filename, ext = os.path.splitext(filename)
	
	# check if its a npy file and if the name matches the name structure we are expecting
	# if so - its already read and processed so we just return that file name
	if os.path.isfile(filepath) and filename.endswith('_mem_check_cache_' + read_mode):
		return filepath
	
	# filepath is not a _mem_check_cache.npy file.
	# calculate the sha1 function and check if a cache file already exist for that file
	_sha1 = calc_sha1(filepath)
	cache_file_name = filename + '_' + _sha1 + '_mem_check_cache_' + read_mode
	
	if save_path:
		cache_full_path = os.path.join(save_path, cache_file_name)
	else:
		cache_full_path = os.path.join(path, cache_file_name)

	if os.path.isfile(cache_full_path+'.npy'):
		return cache_full_path+'.npy'
	
	fid = open(filepath, 'r')
	meta = fromstring(fid.readline().strip(), dtype=int, sep='\t')
	
	pagesPerBlock = meta[2]
	bytesPerPage = meta[3]
	architecture = meta[6]
	
	pages_order = fromstring(fid.readline().strip(), dtype=int, sep='\t')
	
	if architecture == 0:  # mlc
		summed_errors = zeros((int(pagesPerBlock / 2), bytesPerPage * 8), 'uint32')
		pages_coupling = pages_order.reshape((int(pagesPerBlock / 2), 2), order='F')
	else:  # tlc
		summed_errors = zeros((int(pagesPerBlock / 3), bytesPerPage * 8), 'uint32')
		pages_coupling = pages_order.reshape((int(pagesPerBlock / 3), 3), order='F')
	
	m = zeros((pagesPerBlock, bytesPerPage * 8), 'uint32')
	
	line = 'dummy string'
	while line:
		line = fid.readline()
		if read_mode == 'normal':
			# in normal read mode we need to reset m each line that we read in order to be able to detect
			# if a cell error caused bit error  in more then one page.
			m = zeros((pagesPerBlock, bytesPerPage * 8), 'uint32')
			
		arr = fromstring(line.strip(), dtype=uint32, sep='\t')
		i = 1
		length = arr.shape[0]
		
		while i < length:
			page = arr[i]
			i = i + 1
			bits = arr[i + 1:i + arr[i] + 1]
			i = i + arr[i] + 1
			bits = concatenate((bits, arr[i + 1:i + arr[i] + 1]))
			i = i + arr[i] + 1
			if read_mode == 'normal':
				m[page, bits] = 1
			elif read_mode == 'levels':
				m[page, bits] += 1
			else:
				raise NotImplemented
		
		if read_mode == 'normal':
			summed_errors_per_loop = zeros(summed_errors.shape, 'uint32')
			i = 0
			while i < pages_coupling.shape[0]:
				summed_errors_per_loop[i, :] = (m[pages_coupling[i, :], :]).sum(axis=0)
				i += 1

			summed_errors_per_loop[(summed_errors_per_loop > 1)] = 1
			summed_errors += summed_errors_per_loop
		
	fid.close()
	if read_mode == 'normal':
			left = summed_errors[0::2, :]
			right = summed_errors[1::2, :]
	else:
		if architecture == 0:
			left_low = m[pages_order[0:int(pagesPerBlock/2):2], :]
			right_low = m[pages_order[1:int(pagesPerBlock/2):2], :]
			left_high = m[pages_order[int(pagesPerBlock/2)::2], :]
			right_high = m[pages_order[int(pagesPerBlock/2)+1::2], :]
			
			left = np.concatenate((left_low,left_high))
			right = np.concatenate((right_low, right_high))
		else:
			left_low = m[pages_order[0:int(pagesPerBlock/3):2], :]
			right_low = m[pages_order[1:int(pagesPerBlock/3):2], :]
			left_middle = m[pages_order[int(pagesPerBlock/3):2*int(pagesPerBlock/3):2], :]
			right_middle = m[pages_order[int(pagesPerBlock/3)+1:2*int(pagesPerBlock/3):2], :]
			left_high = m[pages_order[2*int(pagesPerBlock/3)::2], :]
			right_high = m[pages_order[2*int(pagesPerBlock/3)+1::2], :]
			
			left = np.concatenate((left_low, left_middle, left_high))
			right = np.concatenate((right_low, right_middle, right_high))
		
	np.save(cache_full_path, np.concatenate((left, right), axis=1))
		
	return cache_full_path + '.npy'
